Sources/SwiftKuery/ConnectionPoolConnection.swift

```diff
@@ -269,75 +495,204 @@ public class ConnectionPoolConnection: Connection {
+        connection.rollback(to: savepoint) { result in
+            self.runCompletionHandlerKeepingSelfAlive(result: result, onCompletion: onCompletion)
+        }
     }
-    
+
+    /// Rollback the current transaction to the specified savepoint.
+    ///
+    /// - Parameter to savepoint: The name of the savepoint to rollback to.
+    /// - Returns: A QueryResult representing the result of the operation
+    public func rollbackSync(to savepoint: String) -> QueryResult {
+        var result: QueryResult?
+        let semaphore = DispatchSemaphore(value: 0)
+        rollback(to: savepoint) { res in
+            result = res
+            semaphore.signal()
+        }
+        semaphore.wait()
+        guard let resultUnwrapped = result else {
+            return .error(QueryError.noResult("No QueryResult from execute"))
+        }
+        return resultUnwrapped
+    }
+
     /// Release a savepoint.
     ///
     /// - Parameter savepoint: The name of the savepoint to release.
     /// - Parameter onCompletion: The function to be called when the execution of the release savepoint command has completed.
     public func release(savepoint: String, onCompletion: @escaping ((QueryResult) -> ())) {
-        guard let connection = connection else {
+        guard let connection = self.connection else {
             onCompletion(.error(QueryError.connection("Connection is disconnected")))
             return
         }
-        connection.release(savepoint: savepoint, onCompletion: onCompletion)
+        connection.release(savepoint: savepoint) { result in
+            self.runCompletionHandlerKeepingSelfAlive(result: result, onCompletion: onCompletion)
+        }
+    }
+
+    /// Release a savepoint.
+    ///
+    /// - Parameter savepoint: The name of the savepoint to release.
+    /// - Returns: A QueryResult representing the result of the operation
+    public func releaseSync(savepoint: String) -> QueryResult {
+        var result: QueryResult?
+        let semaphore = DispatchSemaphore(value: 0)
+        release(savepoint: savepoint) { res in
+            result = res
+            semaphore.signal()
+        }
+        semaphore.wait()
+        guard let resultUnwrapped = result else {
+            return .error(QueryError.noResult("No QueryResult from execute"))
+        }
+        return resultUnwrapped
+    }
+
+    // Offload the passed closure keeping self in scope
+    // By calling this function to offload the passed closure we keep the connection wrapper in scope and prevent it being returned to the pool early by virtue of the fact self has to be captured in the calling closure.
+    // The funxction also stores a reference to the wrapper on any ResultSet that is being returned which prevents a connection being returned to the pool until ResultSet.done() is called.
+    private func runCompletionHandlerKeepingSelfAlive(result: QueryResult, onCompletion: @escaping ((QueryResult) -> ())) {
```

Comment:
I need to think long and hard about what's going on here! I see what you're doing...

Reply:
Update name following further consideration.